// gerencianet/app.js
const axios = require('axios');
const fs = require('fs');
const path = require('path');
const https = require('https');

const authenticate = ({ clientID, clientSecret, entrypoint, certificate }) => {

    const cert = fs.readFileSync(certificate)

    const agent = new https.Agent({
        pfx: cert,
        passphrase: ''
    });
    const credentials = Buffer.from(
        `${clientID}:${clientSecret}`
    ).toString('base64');

    return axios({
        method: 'POST',
        url: `${entrypoint}/oauth/token`,
        headers: {
            Authorization: `Basic ${credentials}`,
            'Content-Type': 'application/json'
        },
        httpsAgent: agent,
        data: {
            grant_type: 'client_credentials'
        }
    });
};

const GNRequest = async (credentials) => {
    const authResponse = await authenticate(credentials);
    const accessToken = authResponse.data?.access_token;

    const cert = fs.readFileSync(credentials.certificate)

    const agent = new https.Agent({
        pfx: cert,
        passphrase: ''
    });

    return axios.create({
        baseURL: `${credentials.entrypoint}`,
        httpsAgent: agent,
        headers: {
            Authorization: `Bearer ${accessToken}`,
            'Content-Type': 'application/json'
        }
    });
}

module.exports = GNRequest;
